src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -42,471 +44,524 @@ public static class SystemNamespaceType
                 .WithReturnResultBuilder(ItemsResultBuilder, GetItemsReturnType(LanguageConstants.String, LanguageConstants.Any))
-                .Build(),
-        }.ToImmutableArray();
+                .Build();
+
+            if (featureProvider.LambdasEnabled)
+            {
+                yield return new FunctionOverloadBuilder("flatten")
+                    .WithGenericDescription("Flattens an array of arrays into an array containing all of the individual array elements in the original order.")
+                    .WithVariableParameter("array", LanguageConstants.Array, 0, "TBD")
+                    .WithReturnType(LanguageConstants.Array)
+                    .Build();
+
+                yield return new FunctionOverloadBuilder("filter")
+                    .WithGenericDescription("Filters an array with a custom filtering function.")
+                    .WithRequiredParameter("array", LanguageConstants.Array, "The array to filter")
+                    .WithRequiredParameter("lambda", OneParamLambda(LanguageConstants.Any, LanguageConstants.Bool), "The filtering function",
+                        calculator: getArgumentType => CalculateLambdaFromArrayParam(getArgumentType, 0, t => OneParamLambda(t, LanguageConstants.Bool)))
+                    .WithReturnResultBuilder((binder, fileResolver, diagnostics, arguments, argumentTypes) => {
+                        return new(argumentTypes[0]);
+                    }, LanguageConstants.Array)
+                    .Build();
+
+                yield return new FunctionOverloadBuilder("map")
+                    .WithGenericDescription("Applies a custom mapping function to each element of an array and returns the result array.")
+                    .WithRequiredParameter("array", LanguageConstants.Array, "The array to map")
+                    .WithRequiredParameter("lambda", OneParamLambda(LanguageConstants.Any, LanguageConstants.Any), "The mapping function",
+                        calculator: getArgumentType => CalculateLambdaFromArrayParam(getArgumentType, 0, t => OneParamLambda(t, LanguageConstants.Any)))
+                    .WithReturnResultBuilder((binder, fileResolver, diagnostics, arguments, argumentTypes) => argumentTypes[1] switch {
+                        LambdaType lambdaType => new(new TypedArrayType(lambdaType.ReturnType.Type, TypeSymbolValidationFlags.Default)),
+                        _ => new(LanguageConstants.Any),
+                    }, LanguageConstants.Array)
+                    .Build();
+
+                yield return new FunctionOverloadBuilder("sort")
+                    .WithGenericDescription("Sorts an array with a custom sort function.")
+                    .WithRequiredParameter("array", LanguageConstants.Array, "The array to sort")
+                    .WithRequiredParameter("lambda", TwoParamLambda(LanguageConstants.Any, LanguageConstants.Any, LanguageConstants.Bool), "The sorting function",
```

Comment:
Not related to this PR, but is there a single-argument `sort` override? Would be useful for arrays of strings or ints, which would I guess require an identity lambda (`sort(arrayArg, i => i)`) otherwise?

Reply:
There is not, but we do have a path forward for adding it in the future! I think chances are there are a few other functions we'll want to implement - Marcin suggested a `.ToDictionary(..)` equivalent for example.